src/main/java/bio/terra/workspace/db/WorkspaceDao.java

```diff
@@ -35,7 +37,12 @@ public String createWorkspace(UUID workspaceId, JsonNullable<UUID> spendProfile)
     paramMap.put("spend_profile", spendProfile.orElse(null));
     paramMap.put("spend_profile_settable", !spendProfile.isPresent());
 
-    jdbcTemplate.update(sql, paramMap);
+    try {
+      jdbcTemplate.update(sql, paramMap);
+    } catch (DuplicateKeyException e) {
+      throw new DuplicateWorkspaceException(
+          "Workspace " + workspaceId.toString() + " already exists.", e);
+    }
```

Comment:
could you add a unit test for this? I know it's a bunch more code. It would be good to see that we return a nicely-worded error message (without raw SQL!) in this case.

Reply:
Done, added one at both DAO and service-level.

Also, the alternating messages was caused by Stairway failing to undo steps. You'd get something like this:

Run 1:
1. Create workspace with ID in the WM's DB. (succeeds)
2. Create workspace IAM object in Sam (fails because there's no token)
3. Try to undo the create-in-Sam step by deleting in Sam (also fails because there's still no token). This gives the "no token" error.
4. Because there was a failed undo step, Stairway can't undo any more, so it throws up its hands and calls the flight done. The workspace is never cleaned up from the DB.

Run 2:
1. Create workspace with ID in the WM's DB (fails, since it already exists from Run 1). This gives the "workspace already exists" error.
2. Undo the create-workspace step by deleting the workspace (succeeds). Because the workspace has been cleaned up, the next call will be Run 1 again.

Side note, this meant that unauthenticated users had the ability to delete any workspace by calling "create" on it when it already exists. That's... a little terrifying, and I want to think about our other Flights to make sure this kind of bug (accidentally changing state by "undoing" differently than "doing") doesn't happen elsewhere.